File: easy/1114_print_in_order/solution.cpp
#include <functional>

class Foo {
private:
  std::condition_variable cv;
  std::mutex m;

  uint state;

public:
  Foo() { state = 0; }

  void first(std::function<void()> printFirst) {
    {
      auto lock = std::lock_guard{m};
      printFirst();
      state = 1;
    }
    cv.notify_all();
  }

  void second(std::function<void()> printSecond) {
    {
      auto lock = std::unique_lock{m};
      cv.wait(lock, [&]() { return state == 1; });
      printSecond();
      state = 2;
    }
    cv.notify_all();
  }

  void third(std::function<void()> printThird) {
    auto lock = std::unique_lock{m};
    cv.wait(lock, [&]() { return state == 2; });

    printThird();
    state = 3;
  }
};

File: easy/1114_print_in_order/solution.py
from typing import Callable


class Foo:
    def __init__(self):
        self.called = [0] * 3

    def first(self, printFirst: "Callable[[], None]") -> None:
        printFirst()
        self.called[0] = 1

    def second(self, printSecond: "Callable[[], None]") -> None:
        while not self.called[0]:
            pass

        printSecond()
        self.called[1] = 1

    def third(self, printThird: "Callable[[], None]") -> None:
        while not self.called[1]:
            pass

        printThird()
        self.called[2] = 1
